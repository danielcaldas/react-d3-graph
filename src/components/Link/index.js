import React from 'react';

export default class Link extends React.Component {
    // Properties more likely to mutate are evaluated first to take advantage of short-circuit evaluation
    shouldComponentUpdate(nextProps) {
        return nextProps.x1 !== this.props.x1 || nextProps.y1 !== this.props.y1
            || nextProps.x2 !== this.props.x2 || nextProps.y2 !== this.props.y2
            || nextProps.opacity !== this.props.opacity
            || nextProps.stroke !== this.props.stroke
            || nextProps.strokeWidth !== this.props.strokeWidth
            || nextProps.color !== this.props.color
            || nextProps.semanticStrokeWidth !== this.props.semanticStrokeWidth;
    }

    /**
     * Handle link click event.
     * @return {undefined}
     */
    handleOnClickLink = () => this.props.onClickLink && this.props.onClickLink(this.props.source, this.props.target);

    render() {
        const lineStyle = {
            strokeWidth: this.props.strokeWidth,
            stroke: this.props.stroke,
            opacity: this.props.opacity
        };

        const lineProps = {
            className: this.props.className,
            onClick: this.handleOnClickLink,
            style: lineStyle,
            x1: this.props.x1,
            x2: this.props.x2,
            y1: this.props.y1,
            y2: this.props.y2
        };

        return (
            <line {...lineProps}/>
        );
    }
}
