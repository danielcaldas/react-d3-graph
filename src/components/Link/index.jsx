import React from 'react';

/**
 * Link component is responsible for encapsulating link render.
 * @example
 * const onClickLink = function(source, target) {
 *      window.alert(`Clicked link between ${source} and ${target}`);
 * };
 *
 * <Link
 *     source='idSourceNode'
 *     target='idTargetNode'
 *     x1=22
 *     y1=22
 *     x2=22
 *     y2=22
 *     strokeWidth=1.5
 *     stroke='green'
 *     className='link'
 *     opacity=1
 *     onClickLink={onClickLink} />
 */
export default class Link extends React.Component {
    /**
     * Handle link click event.
     * @return {undefined}
     */
    handleOnClickLink = () => this.props.onClickLink && this.props.onClickLink(this.props.source, this.props.target);

    shouldComponentUpdate(nextProps) {
        // Properties more likely to mutate are evaluated first to take advantage of short-circuit evaluation
        return nextProps.x1 !== this.props.x1 || nextProps.y1 !== this.props.y1
            || nextProps.x2 !== this.props.x2 || nextProps.y2 !== this.props.y2
            || nextProps.opacity !== this.props.opacity
            || nextProps.stroke !== this.props.stroke
            || nextProps.strokeWidth !== this.props.strokeWidth
            || nextProps.color !== this.props.color;
    }

    render() {
        const lineStyle = {
            strokeWidth: this.props.strokeWidth,
            stroke: this.props.stroke,
            opacity: this.props.opacity
        };

        const lineProps = {
            className: this.props.className,
            onClick: this.handleOnClickLink,
            style: lineStyle,
            x1: this.props.x1,
            x2: this.props.x2,
            y1: this.props.y1,
            y2: this.props.y2
        };

        return (
            <line {...lineProps}/>
        );
    }
}
