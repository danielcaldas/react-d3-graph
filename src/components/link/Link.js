import React from "react";

/**
 * Link component is responsible for encapsulating link render.
 * @example
 * const onClickLink = function(source, target) {
 *      window.alert(`Clicked link between ${source} and ${target}`);
 * };
 *
 * const onRightClickLink = function(source, target) {
 *      window.alert(`Right clicked link between ${source} and ${target}`);
 * };
 *
 * const onMouseOverLink = function(source, target) {
 *      window.alert(`Mouse over in link between ${source} and ${target}`);
 * };
 *
 * const onMouseOutLink = function(source, target) {
 *      window.alert(`Mouse out link between ${source} and ${target}`);
 * };
 *
 * <Link
 *     d="M1..."
 *     source="idSourceNode"
 *     target="idTargetNode"
 *     markerId="marker-small"
 *     strokeWidth=1.5
 *     stroke="green"
 *     className="link"
 *     opacity=1
 *     mouseCursor="pointer"
 *     onClickLink={onClickLink}
 *     onRightClickLink={onRightClickLink}
 *     onMouseOverLink={onMouseOverLink}
 *     onMouseOutLink={onMouseOutLink} />
 */
export default class Link extends React.Component {
    /**
     * Handle link click event.
     * @returns {undefined}
     */
    handleOnClickLink = () => this.props.onClickLink && this.props.onClickLink(this.props.source, this.props.target);

    /**
     * Handle link right click event.
     * @param {Object} event - native event.
     * @returns {undefined}
     */
    handleOnRightClickLink = event =>
        this.props.onRightClickLink && this.props.onRightClickLink(event, this.props.source, this.props.target);

    /**
     * Handle mouse over link event.
     * @returns {undefined}
     */
    handleOnMouseOverLink = () =>
        this.props.onMouseOverLink && this.props.onMouseOverLink(this.props.source, this.props.target);

    /**
     * Handle mouse out link event.
     * @returns {undefined}
     */
    handleOnMouseOutLink = () =>
        this.props.onMouseOutLink && this.props.onMouseOutLink(this.props.source, this.props.target);

    render() {
        const lineStyle = {
            strokeWidth: this.props.strokeWidth,
            stroke: this.props.stroke,
            opacity: this.props.opacity,
            fill: "none",
            cursor: this.props.mouseCursor,
        };

        const lineProps = {
            className: this.props.className,
            d: this.props.d,
            onClick: this.handleOnClickLink,
            onContextMenu: this.handleOnRightClickLink,
            onMouseOut: this.handleOnMouseOutLink,
            onMouseOver: this.handleOnMouseOverLink,
            style: lineStyle,
        };

        if (this.props.markerId) {
            lineProps.markerEnd = `url(#${this.props.markerId})`;
        }

        const { label, id } = this.props;
        const textProps = {
            dy: -1,
            style: {
                fill: this.props.fontColor,
                fontSize: this.props.fontSize,
                fontWeight: this.props.fontWeight,
            },
        };

        return (
            <g>
                <path {...lineProps} id={id} />
                {label && (
                    <text style={{ textAnchor: "middle" }} {...textProps}>
                        <textPath href={`#${id}`} startOffset="50%">
                            {label}
                        </textPath>
                    </text>
                )}
            </g>
        );
    }
}
