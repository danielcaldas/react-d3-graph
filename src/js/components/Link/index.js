import React from 'react';

export default class Link extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            source: this.props.link.source,
            target: this.props.link.target,
            x1: this.props.link.x1,
            x2: this.props.link.x2,
            y1: this.props.link.y1,
            y2: this.props.link.y2,
            onClickLink: this.props.onClickLink
        };
    }

    componentWillReceiveProps(nextProps) {
        const update = nextProps.link.x1 !== this.state.x1 || nextProps.link.y1 !== this.state.y1
                    || nextProps.link.x2 !== this.state.x2 || nextProps.link.y2 !== this.state.y2;

        if (update) {
            this.setState({
                x1: nextProps.link.x1,
                y1: nextProps.link.y1,
                x2: nextProps.link.x2,
                y2: nextProps.link.y2,
            });
        }
    }

    /**
     * Handle click action in some link between two nodes. It calls a callback passed by parent component
     * and it passes two parameters, the identifiers of the source and target nodes in the relation.
     * @return {undefined}
     */
    handleOnClickLink = () => {
        if (this.state.onClickLink) {
            this.state.onClickLink(this.state.source, this.state.target);
        }
    }

    render() {
        const lineStyle = {
            strokeWidth: '1.5',
            stroke: 'rgb(169, 169, 169)',
            opacity: 1
        };

        const lineProps = {
            className: 'link',
            onClick: this.handleOnClickLink,
            style: lineStyle,
            x1: this.state.x1,
            x2: this.state.x2,
            y1: this.state.y1,
            y2: this.state.y2
        };

        return (
            <line {...lineProps}/>
        );
    }
}
