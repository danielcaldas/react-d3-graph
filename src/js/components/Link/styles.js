export default {
    lineStyle: {
        strokeWidth: '1.5',
        stroke: 'rgb(169, 169, 169)',
        opacity: 1
    }
};
