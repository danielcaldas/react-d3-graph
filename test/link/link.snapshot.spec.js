import React from "react";
import renderer from "react-test-renderer";

import Link from "../../src/components/link/Link";

describe("Snapshot - Link Component", () => {
  let that = {};

  beforeEach(() => {
    that.callbackMock = jest.fn();

    that.link = renderer.create(
      <Link
        x1="2"
        y1="2"
        x2="4"
        y2="4"
        opacity="1"
        stroke="red"
        strokeWidth="2"
        onClickLink={that.callbackMock}
        strokeDasharray={0}
        strokeDashoffset={0}
        strokeLinecap="butt"
      />
    );

    that.tree = that.link.toJSON();

    const viewGenerator = () => () => "viewGenerator";
    that.linkForViewGenerator = renderer.create(
      <Link
        x1="2"
        y1="2"
        x2="4"
        y2="4"
        opacity="1"
        stroke="red"
        strokeWidth="2"
        onClickLink={that.callbackMock}
        strokeDasharray={0}
        strokeDashoffset={0}
        strokeLinecap="butt"
        node={{ source: { x: 2, y: 2 }, target: { x: 4, y: 4 } }}
        viewGenerator={viewGenerator}
      />
    );

    that.treeForViewGenerator = that.linkForViewGenerator.toJSON();
  });

  test("should match snapshot", () => {
    expect(that.tree).toMatchSnapshot();
  });
  test("should match snapshot for viewGenerator", () => {
    expect(that.treeForViewGenerator).toMatchSnapshot();
  });
});
