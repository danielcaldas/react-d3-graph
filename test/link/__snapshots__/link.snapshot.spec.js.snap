// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot - Link Component should match snapshot 1`] = `
<svg>
  <path
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    style={
      Object {
        "cursor": undefined,
        "fill": "none",
        "opacity": "1",
        "stroke": "red",
        "strokeWidth": "2",
      }
    }
  />
</svg>
`;
